cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh

```diff
@@ -584,22 +666,27 @@ void train_per_cluster(const handle_t& handle,
       rot_vectors.data(),
       pq_n_rows,
       index.pq_book_size(),
-      index.pq_centers().data_handle() + index.pq_book_size() * index.pq_len() * l,
+      pq_centers_tmp.data() + size_t(index.pq_book_size()) * size_t(index.pq_len()) * size_t(l),
       pq_labels.data(),
       pq_cluster_sizes.data(),
       raft::distance::DistanceType::L2Expanded,
       stream,
       device_memory);
   }
+  transpose_pq_centers(index, pq_centers_tmp.data(), stream);
 }
 
-/** See raft::spatial::knn::ivf_pq::extend docs */
+/**
+ * See raft::spatial::knn::ivf_pq::extend docs.
+ *
+ * This version requires `new_vectors` and `new_indices` (if non-null) to be on-device.
+ */
 template <typename T, typename IdxT>
-inline auto extend(const handle_t& handle,
-                   const index<IdxT>& orig_index,
-                   const T* new_vectors,
-                   const IdxT* new_indices,
-                   IdxT n_rows) -> index<IdxT>
+inline auto extend_device(const handle_t& handle,
```

Comment:
While I don't mind this for now on the pointer-based API (we're going to be eventually deprecating the pointer-based APIs altogether in favor of `mdspan`), I think the mdspan-based API is really going to clean this up because we'll have a single `extend()` function that accepts both device and host mdspan (and managed mdspan), so we can either add overloads or dispatch based on the accessor type.

Reply:
Created a task: https://github.com/rapidsai/raft/issues/711 (11)